target/scripts/helper-functions.sh

```diff
@@ -150,51 +150,48 @@ export -f _sanitize_ipv4_to_subnet_cidr
 
 function _extract_certs_from_acme
 {
-  local KEY
-  # shellcheck disable=SC2002
-  KEY=$(cat /etc/letsencrypt/acme.json | python -c "
-import sys,json
-acme = json.load(sys.stdin)
-for key, value in acme.items():
-    certs = value['Certificates']
-    if certs is not None:
-        for cert in certs:
-            if 'domain' in cert and 'key' in cert:
-                if 'main' in cert['domain'] and cert['domain']['main'] == '${1}' or 'sans' in cert['domain'] and '${1}' in cert['domain']['sans']:
-                    print cert['key']
-                    break
-")
-
-  local CERT
-  # shellcheck disable=SC2002
-  CERT=$(cat /etc/letsencrypt/acme.json | python -c "
-import sys,json
-acme = json.load(sys.stdin)
-for key, value in acme.items():
-    certs = value['Certificates']
-    if certs is not None:
-        for cert in certs:
-            if 'domain' in cert and 'certificate' in cert:
-                if 'main' in cert['domain'] and cert['domain']['main'] == '${1}' or 'sans' in cert['domain'] and '${1}' in cert['domain']['sans']:
-                    print cert['certificate']
-                    break
-")
-
-  if [[ -n "${KEY}${CERT}" ]]
+  local CERT_DOMAIN=${1}
+  if [[ -z ${CERT_DOMAIN} ]]
   then
-    mkdir -p "/etc/letsencrypt/live/${HOSTNAME}/"
+    _notify 'err' "_extract_certs_from_acme | CERT_DOMAIN is empty"
+    return 1
+  fi
 
-    echo "${KEY}" | base64 -d >/etc/letsencrypt/live/"${HOSTNAME}"/key.pem || exit 1
-    echo "${CERT}" | base64 -d >/etc/letsencrypt/live/"${HOSTNAME}"/fullchain.pem || exit 1
-    _notify 'inf' "Cert found in /etc/letsencrypt/acme.json for ${1}"
+  local KEY CERT
+  KEY=$(acme_extract /etc/letsencrypt/acme.json "${CERT_DOMAIN}" --key)
+  CERT=$(acme_extract /etc/letsencrypt/acme.json "${CERT_DOMAIN}" --cert)
 
-    return 0
-  else
+  if [[ -z ${KEY} || -z ${CERT} ]]
+  then
+    _notify 'warn' "_extract_certs_from_acme | Unable to find key & cert for '${CERT_DOMAIN}' in '/etc/letsencrypt/acme.json'"
     return 1
   fi
+
+  # Currently we advise SSL_DOMAIN for wildcard support using a `*.example.com` value,
+  # The filepath however should be `example.com`, avoiding the wildcard part:
+  if [[ ${SSL_DOMAIN} == "${CERT_DOMAIN}" ]]
+  then
+    CERT_DOMAIN=$(_strip_wildcard_prefix "${SSL_DOMAIN}")
+  fi
+
+  mkdir -p "/etc/letsencrypt/live/${CERT_DOMAIN}/"
+  echo "${KEY}" | base64 -d > "/etc/letsencrypt/live/${CERT_DOMAIN}/key.pem" || exit 1
```

Comment:
Just a question, if that is the intended behaviour: When this command fails, then check-for-changes should exit with errorcode 1. Is that correct?
Supervisor will then start check-for-changes again.

Reply:
I didn't author this line originally, but I guess that's what happens? :sweat_smile: 

I'm not quite sure how it'd fail if the directory was created and the value is not empty, other than maybe not being valid `base64` encoded data for some reason?

It would end up in a bit of a loop then (_startup script also calls this method in `setup-stack.sh`_), without an error logged? (or that comes from the base64 decode error prior to exit?

Reply:
> I'm not quite sure how it'd fail

Thought the same. But was wondering why there is this explicit `|| exit 1`

Reply:
> without an error logged?

When supervisor restarts a service, you'll see something similar to this at STDOUT:

```
2021-11-01 23:13:08,348 INFO exited: update-check (terminated by SIGTERM; not expected)
2021-11-01 23:13:08,350 INFO spawned: 'update-check' with pid 3534
2021-11-01 23:13:08,436 INFO success: update-check entered RUNNING state, process has stayed up for > than 0 seconds (startsecs)
```

Reply:
> But was wondering why there is this explicit `|| exit 1`

It was introduced in July 2020 via [this PR](https://github.com/docker-mailserver/docker-mailserver/pull/1553), there is no discussion about it or mention in the relevant commit message.